aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+    )) {
+      done_searching = true;
+      *found = true;
+    } else {
+      left_ind = mid_ind+1;
+    }
+
+    if (left_ind > right_ind) {
+      done_searching = true;
+      *found = false;
+      mid_ind = -1;
+    }
+  }
+
+  return mid_ind;
+}
+
+Tensor& bmm_out_sparse_cpu(Tensor& result, const SparseTensor& self, const Tensor& mat2) {
+  TORCH_CHECK(!mat2.is_sparse(), "bmm_sparse: Tensor 'mat2' must be dense");
+
+  TORCH_CHECK(self.dense_dim() == 0, "bmm_sparse: Tensor 'self' must have 0 dense dims, but has ", self.dense_dim());
+  TORCH_CHECK(self.sparse_dim() == 3, "bmm_sparse: Tensor 'self' must have 3 sparse dims, but has ", self.sparse_dim());
+  TORCH_CHECK(mat2.dim() == 3, "bmm_sparse: Tensor 'mat2' must have 3 dims, but has ", mat2.dim());
+
+  TORCH_CHECK(self.size(0) == mat2.size(0), "bmm_sparse: 'self.size(0)' and 'mat2.size(0)' must match");
+  TORCH_CHECK(self.size(2) == mat2.size(1), "bmm_sparse: 'self.size(2)' and 'mat2.size(1)' must match");
+
+  result.resize_({self.size(0), self.size(1), mat2.size(2)});
+
+  if (self._nnz() == 0) {
+    result.zero_();
+    return result;
+  }
+
+  // First need to coalesce to get all of the first dimension indices
+  // in order since we'll be sending each matrix into the MM operation
+  SparseTensor self_coalesced;
+
+  if (!self.is_coalesced()){
+    self_coalesced = self.coalesce();
+  } else {
+    self_coalesced = self;
+  }
+
+  int64_t nnz =        self_coalesced._nnz();
+  LongTensor indices = self_coalesced._indices();
+  Tensor values =      self_coalesced._values();
+
+  LongTensor indices_dim0 = indices[0];
+  int64_t indices_dim0_stride = indices_dim0.stride(0);
+  int64_t* indices_dim0_ptr = indices_dim0.data_ptr<int64_t>();
+  LongTensor indices_dim1_dim2 = indices.slice(0, 1, 3);
+
+  int64_t dim_i = self_coalesced.size(1);
+  int64_t dim_j = self_coalesced.size(2);
+  int64_t dim_k = mat2.size(2);
+
+  Scalar beta = 0;
+  Tensor t = at::ones({self_coalesced.size(1)});
+  Scalar alpha = 1;
+
+  int64_t mat_el_begin_idx = 0;
+
+  int64_t num_matrices = self_coalesced.size(0);
+
+  // Iterate through each set of 2D matrices within the 3D
+  // tensor inputs, performing a matrix multiply with each one.
+  int64_t start_mat_num = indices_dim0_ptr[0];
+  for (int64_t cur_mat_num = 0;
+    (cur_mat_num < num_matrices);
+    cur_mat_num++
+  ) {
+    // If there are sparse matrices at the beginning or end that
+    // have all zero elements, we need to zero out the result matrix.
+    if ((cur_mat_num < start_mat_num) || (mat_el_begin_idx >= nnz)) {
+      result[cur_mat_num].zero_();
+      continue;
+    }
+
+    // Search for the range of sparse tensor elements that
+    // correspond to the current matrix number. We already know
+    // where the current matrix begins, so we just need to find
+    // the end. The search excludes everything to the left of
+    // the starting point, for best performance
```

Comment:
Don't have to do it in this PR, but since you need to find *all* matix splits, I wonder if it wouldn't be better to do a single computation that precomputes all of the splits into a buffer, instead of repeatedly binary searching inside the batch loop.

Reply:
If there's a performant way to do that, sure. Right now, the performance of finding the matrix splits is on the order of `O(num_matrices * ln(num_matrices * nnz_per_matrix))` (actually a bit better than that, since we're starting each subsequent search from the end of the previous search result). If we can get something better by somehow combining all the searches into one, then we should do that. But I don't really have a great idea how at the moment.

Reply:
Well, the shitty algorithm is to just iterate over every entry one by one and record a split whenever it happens. That's `O(nnz)` but you have much better cache locality.

Reply:
Linear iteration might not be so bad depending on the circumstance. It depends on how much faster (in terms of latency per access) that a linear access pattern is compared to a binary search access pattern, the caller's ratio of nonzeros versus number of matrices, and the total number of nonzeros.

At my previous job where I did competitive analysis of cache/memory performance on smartphone CPUs, it was common for a random memory or L3 cache access latency to be around 4-12 times slower than linear access latency (here's a memory latency graph of the Apple A13: https://www.anandtech.com/show/14892/the-apple-iphone-11-pro-and-max-review/3 compare the "linear chain" and "full random" lines on the chart), since prefetchers are good at handling linearity. I think we can say that a binary search's access pattern will behave fairly similarly to a random access (although maybe a little bit lower latency) as far as the prefetcher is concerned. When you get down to L2 the latency is about twice as fast for linear accesses versus random, and in L1, the access pattern makes no difference at all.

So a linear search will tend to be faster than the binary search if the total number of accesses multiplied by the access pattern latency is lower than that of the binary search. If we look at the best possible case where linear access could win, we'd look at the case where our total number of non-zeros is too large to fit within L2 and each linear access is 12 times faster than each binary search access. Applying this factor to the big-O runtimes of the two methods, we'd say that the linear search is preferable if `N < 12 * M * log2(N)`, where N is the total number of non-zeros and M is the number of matrices. If our L2 is 4 MB and we're using 32-bit floats, then N must be significantly more than 1,000,000 to get the 12x linear speedup--let's say we need `N > 1,000,000`. If we plug this into Wolfram Alpha: `N < 12 * M * log2(N), where 1000000 < N < 10000000 and 1 < M < 10000000`, we see that the inequality is true in almost the whole space (paying only attention to the part of the plot where N >= M https://www.wolframalpha.com/input/?i=N+%3C+12+*+M+*+log2%28N%29%2C+where+1000000+%3C+N+%3C+10000000+and+1+%3C+M+%3C+10000000), meaning that the linear access pattern actually could be faster in most cases (binary search is better if N/M is more than about 240). Even if the linear pattern does not give us any speedup, the binary search is only useful if the ratio of N/M is more than about 20: https://www.wolframalpha.com/input/?i=N+%3C+1+*+M+*+log2%28N%29%2C+where+1+%3C+N+%3C+100000+and+1+%3C+M+%3C+100000

But I don't know what a typical N/M ratio would be, so it's hard to know for sure if my reasoning here is realistic.

Reply:
For what it's worth, I tried running a linear search in the CPU implementation. The performance was barely any different than the binary search:


num_matrices | squ_mat_elements | sparsity | nnz | builtin | builtin-coalesced
-- | -- | -- | -- | -- | --
10 | 100 | 0.9 | 100 | 0.9993763642 | 0.9941944848
10 | 100 | 0 | 1000 | 0.9943479171 | 0.9892473118
10 | 10000 | 0.999 | 100 | 1.000578369 | 0.9982660391
10 | 10000 | 0.99 | 1000 | 0.9908002336 | 0.965602518
10 | 10000 | 0.9 | 10000 | 1.01888459 | 0.9785580659
10 | 10000 | 0 | 100000 | 0.9809957734 | 0.929685379
10 | 1000000 | 0.999 | 10000 | 1.042729942 | 1.019447861
10 | 1000000 | 0.99 | 100000 | 1.025178905 | 1.00487879
100 | 100 | 0.9 | 1000 | 0.9959922814 | 0.9865570492
100 | 100 | 0 | 10000 | 0.9971276159 | 0.9865919064
100 | 10000 | 0.999 | 1000 | 0.9997310579 | 0.9952899695
100 | 10000 | 0.99 | 10000 | 0.9963695038 | 0.9998594881
100 | 10000 | 0.9 | 100000 | 0.9903190667 | 0.9664209664
100 | 1000000 | 0.999 | 100000 | 1.019451 | 1.047661575
1000 | 100 | 0.9 | 10000 | 0.9977529149 | 0.9945412865
1000 | 100 | 0 | 100000 | 0.9855607651 | 0.9915879947
1000 | 10000 | 0.999 | 10000 | 1.002447786 | 1.002210358
1000 | 10000 | 0.99 | 100000 | 1.049603886 | 1.04982573
10000 | 100 | 0.9 | 100000 | 0.9992745346 | 0.9982321546
10000 | 10000 | 0.999 | 100000 | 1.01204083 | 1.011304983

So maybe I should just use a linear search since the code is simpler. But again I don't know if my sizing parameters are realistic or not. What do you think?